python/src/pyarrow/adapters/pandas.cc

```diff
@@ -706,8 +735,10 @@ Status ArrowToPandas(const std::shared_ptr<Column>& col, PyObject** out) {
     FROM_ARROW_CASE(DOUBLE);
     FROM_ARROW_CASE(STRING);
     default:
+      Py_DECREF(py_ref);
       return Status::NotImplemented("Arrow type reading not implemented");
   }
+  Py_DECREF(py_ref);
```

Comment:
I wonder (out loud) if it possible to use `OwnedRef` / RAII-like semantics to reduce some of the manual reference counting (especially the error cases).